packages/loader/container-loader/src/containerStorageAdapter.ts

```diff
@@ -126,12 +128,25 @@ export class ContainerStorageAdapter implements IDocumentStorageService, IDispos
 	}
 
 	public async getSnapshot(snapshotFetchOptions?: ISnapshotFetchOptions): Promise<ISnapshot> {
-		if (this._storageService.getSnapshot !== undefined) {
-			return this._storageService.getSnapshot(snapshotFetchOptions);
+		if (this._storageService.getSnapshot === undefined) {
+			throw new UsageError(
+				"getSnapshot api should exist in internal storage in ContainerStorageAdapter",
+			);
+		}
+
+		const key = snapshotFetchOptions?.loadingGroupIds?.sort().join();
+		const hasLoadingGroupIds = key !== undefined && key !== "";
+		// So when do we drop the this.loadingGroupIds cache? The problem is that we need it if we go offline.
+		const snapshot = hasLoadingGroupIds
+			? this.loadingGroupIds[key]
+			: await this._storageService.getSnapshot(snapshotFetchOptions);
+		if (
+			snapshotFetchOptions?.loadingGroupIds !== undefined &&
+			snapshotFetchOptions.loadingGroupIds.length > 0
+		) {
+			this.loadingGroupIds[snapshotFetchOptions.loadingGroupIds.sort().join()] = snapshot;
```

Comment:
I initially tried to cache a loading group Id to it's call, but that seemed to be especially complex.

For example, if I store individual snapshots containing different groupIds with different sequence numbers, how is that supposed to behave?

Let's say I make a call to groupId1 & groupId2. What should the snapshot format be? Because they are at different sequence numbers, groupId1 needs to process op 51-100 first and groupId2 is already at 100. What sequence number do we put on the ISnapshot format that we return? Or do we error out in this case? 
```
{ 
    groupId1: { seq: 50, snapshot: ... },
    groupId2: { seq: 100, snapshot: ... },
}
```

Reply:
And how do I update this cache if per say, I make a groupId 1 call at seq 50, and then later on I make a groupId 1 & 2 call at seq 100. Then it would be optimal to update groupId 1 from an optimization standpoint.

Reply:
I think we should allow runtime to only ever ask for one groupID.
Driver and protocol, however, should support fetching multiple groupIDs - this will be required for "refresh snapshot" scenarios

Reply:
@tyler-cai-microsoft Is there some plain eng description somewhere that we can read to better understand the logic?

Reply:
No, the logic is scratchwork and is going to be rewritten. Long story short, it's bad and it does not solve the problem enough at hand.